docs/design/coreclr/botr/memory-model.md

```diff
@@ -0,0 +1,248 @@
+
+# CLR memory model
+
+## ECMA vs. CLR memory models.
+ECMA 335 standard defines a very weak memory model. After two decades the desire to have a flexible model did not result in considerable benefits due to hardware being more strict. On the other hand programming against ECMA model requires extra complexity to handle scenarios that are hard to comprehend and not possible to test.
+
+In the course of multiple releases CLR implementation settled around a memory model that is a practical compromise between what can be implemented efficiently on the current hardware, while staying reasonably approachable by the developers. This document rationalizes the invariants provided and expected by the CLR runtime in its current implementation with expectation of that being carried to future releases.
```

Comment:
```suggestion
In the course of multiple releases CLR implementation settled around a memory model that is a practical compromise between what can be implemented efficiently on the current hardware, while staying reasonably approachable by the developers. This document rationalizes the invariants provided and expected by the .NET runtime in its current implementation with expectation of that being carried to future releases.
```

Reply:
I would replace "CLR runtime" with ".NET runtime" everywhere.

(Also, CLR stands for "Common Language Runtime", so "CLR runtime" is "Common Language Runtime runtime".)

Reply:
Is all this applicable to both runtimes?

Reply:
There are more than 2  .Net runtimes. This all applies to `.net framework`, `Mono`, `CoreCLR`, `NativeAOT`. 

There are other implementations (`Unity` , `.net Micro Framework`, some others? . . . ), that I am not very familiar with.
I would expect that this will apply for the most part, but there could be some deviations.

Reply:
Unity uses either Mono 6.13/5.11/2.10 (depending on version) with the Boehm GC (so Mono features should apply to it) or their own IL2CPP runtime that's an AOT and also uses the Boehm GC (but it could differ from the rest of the runtimes in the most visible way.)
Worth noting is the fact that as far as I understand, Unity is going to keep using the IL2CPP runtime even after the porting to .Net 7/8 and CoreCLR, so having the same behaviour with it would be preferred.